L1Trigger/L1TTrackMatch/plugins/L1TkMuonProducer.cc

```diff
@@ -0,0 +1,712 @@
+
+    fIn_theta->Close();
+    fIn_phi->Close();
+
+    // settings : NB : now hardcoded, could be read from cfg
+    mantracorr_ovrl_->set_safety_factor(0.5, 0.5);
+    mantracorr_ovrl_->setArbitrationType("MaxPt");
+  }
+
+  if (emtfMatchAlgoVersion_ == kMantra) {
+    std::string fIn_bounds_name = iConfig.getParameter<edm::FileInPath>("mantra_emtfcorr_boundaries").fullPath();
+    std::string fIn_theta_name = iConfig.getParameter<edm::FileInPath>("mantra_emtfcorr_theta_windows").fullPath();
+    std::string fIn_phi_name = iConfig.getParameter<edm::FileInPath>("mantra_emtfcorr_phi_windows").fullPath();
+
+    auto bounds = L1TkMuMantra::prepare_corr_bounds(fIn_bounds_name, "h_dphi_l");
+
+    TFile* fIn_theta = TFile::Open(fIn_theta_name.c_str());
+    TFile* fIn_phi = TFile::Open(fIn_phi_name.c_str());
+
+    mantracorr_endc_ = std::unique_ptr<L1TkMuMantra>(new L1TkMuMantra(bounds, fIn_theta, fIn_phi, "mantra_endcap"));
+
+    fIn_theta->Close();
+    fIn_phi->Close();
+
+    // settings : NB : now hardcoded, could be read from cfg
+    mantracorr_endc_->set_safety_factor(0.5, 0.5);
+    mantracorr_endc_->setArbitrationType("MaxPt");
+  }
+}
+
+L1TkMuonProducer::~L1TkMuonProducer() {}
+
+// ------------ method called to produce the data  ------------
+void L1TkMuonProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  // the L1Mu objects
+  edm::Handle<RegionalMuonCandBxCollection> l1bmtfH;
+  edm::Handle<RegionalMuonCandBxCollection> l1omtfH;
+  edm::Handle<RegionalMuonCandBxCollection> l1emtfH;
+  edm::Handle<EMTFTrackCollection> l1emtfTCH;
+
+  iEvent.getByToken(bmtfToken, l1bmtfH);
+  iEvent.getByToken(omtfToken, l1omtfH);
+  iEvent.getByToken(emtfToken, l1emtfH);
+  iEvent.getByToken(emtfTCToken, l1emtfTCH);
+
+  // the L1Tracks
+  edm::Handle<L1TTTrackCollectionType> l1tksH;
+  iEvent.getByToken(trackToken, l1tksH);
+
+  TkMuonCollection oc_bmtf_tkmuon;
+  TkMuonCollection oc_omtf_tkmuon;
+  TkMuonCollection oc_emtf_tkmuon;
+
+  std::vector<L1TkMuMantraDF::track_df> mantradf_tracks;  // if needed, just encode once for all trk finders
+  if (bmtfMatchAlgoVersion_ == kMantra || omtfMatchAlgoVersion_ == kMantra || emtfMatchAlgoVersion_ == kMantra) {
+    mantradf_tracks = product_to_trkvec(*l1tksH.product());
+  }
+
+  // process each of the MTF collections separately! -- we don't want to filter the muons
+
+  // ----------------------------------------------------- barrel
+  if (bmtfMatchAlgoVersion_ == kTP)
+    runOnMTFCollection_v1(l1bmtfH, l1tksH, oc_bmtf_tkmuon, 1);
+  else if (bmtfMatchAlgoVersion_ == kMantra) {
+    auto muons = product_to_muvec(*l1bmtfH.product());
+    auto match_idx = mantracorr_barr_->find_match(mantradf_tracks, muons);
+    build_tkMuons_from_idxs(oc_bmtf_tkmuon, match_idx, l1tksH, l1bmtfH, 1);
+  } else
+    throw cms::Exception("TkMuAlgoConfig")
+        << " barrel : trying to run an invalid algorithm (this should never happen)\n";
```

Comment:
useful to print the value of bmtfMatchAlgoVersion_ (in all messages)

Reply:
done